src/VecSim/algorithms/hnsw/hnsw_batch_iterator.cpp

```diff
@@ -0,0 +1,206 @@
+#include "hnsw_batch_iterator.h"
+
+#include <utility>
+#include "VecSim/query_result_struct.h"
+#include "VecSim/utils/vec_utils.h"
+#include "VecSim/algorithms/hnsw/visited_nodes_handler.h"
+
+inline void HNSW_BatchIterator::visitNode(idType node_id) {
+    this->visited_list->tagNode(node_id, this->visited_tag);
+}
+
+inline bool HNSW_BatchIterator::hasVisitedNode(idType node_id) const {
+    return this->visited_list->getNodeTag(node_id) == this->visited_tag;
+}
+
+void HNSW_BatchIterator::prepareResults(
+    VecSimQueryResult_List batch_results,
+    vecsim_stl::max_priority_queue<pair<float, idType>> top_candidates, size_t n_res,
+    VecSimQueryResult_Order order) {
+    size_t initial_results_num = array_len(batch_results);
+    // Put the "spare" results (if exist) in the results heap.
+    while (top_candidates.size() > n_res - initial_results_num) {
+        labelType label = this->index->hnsw.getExternalLabel(top_candidates.top().second);
+        this->results.emplace(top_candidates.top().first, label); // (distance, label)
+        top_candidates.pop();
+    }
+    // Return results from the top candidates heap, put them in reverse order in the batch results
+    // array.
+    for (int i = (int)(top_candidates.size() + initial_results_num - 1);
+         i >= (int)initial_results_num; i--) {
+        array_grow(batch_results);
```

Comment:
you already allocated this array with `n_res` capacity? why the grow?

Reply:
The grow is to indicate that the actual length of the array has increased (not the capacity). This is required since I don't use here `array_append` (as I insert to the array in a specific position...)